src/onnx/parse_multinomial.cpp

```diff
@@ -49,35 +52,95 @@ struct parse_multinomial : op_parser<parse_multinomial>
+        instruction_ref seed_input;
         if(contains(info.attributes, "seed"))
-            gen.seed(info.attributes.at("seed").f());
+        {
+            float seed = info.attributes.at("seed").f();
+            migraphx::shape s{migraphx::shape::float_type, {1}};
+            std::vector<float> data = {seed};
+            seed_input              = info.add_literal(migraphx::literal(s, data));
+        }
+        else
+        {
+            seed_input = info.add_instruction(migraphx::make_op("random_seed"));
+        }
+        instruction_ref randoms;
+
+        shape s0 = args[0]->get_shape();
+
+        if(s0.dynamic())
+        {
+            //  Dynamic batch_size will be taken from args[0].  The input argument to this should
+            // have a second dimension of sample_size.
+            std::vector<shape::dynamic_dimension> dyn_dim_set;
+            dyn_dim_set.emplace_back(s0.dyn_dims().front());
+            dyn_dim_set.emplace_back(shape::dynamic_dimension{sample_size, sample_size});
+
+            // read the input dimensions
+            auto dim_of =
+                info.add_instruction(migraphx::make_op("dimensions_of", {{"end", 2}}), args[0]);
+
+            // The next two operations insert the value sample_size into the second array position
+
+            // make an argument of (1, 0)
+            shape s(shape::int64_type, {2});
+            std::vector<int64_t> data1{1, 0};
+            auto l1        = info.add_literal(s, data1);
+            auto batch_arg = info.add_instruction(migraphx::make_op("mul"), dim_of, l1);
+            std::vector<int64_t> data2(2, 0);
+            // make an argument of (0, sample_size)
+            data2[1]         = sample_size;
+            auto l2          = info.add_literal(s, data2);
+            auto alloc_shape = info.add_instruction(migraphx::make_op("add"), batch_arg, l2);
+            // alloc_shape should contain the input-based shape dimensions as its values at runtime,
+            // and its own shape is (2, 0)
+
+            // compile_shape is the shape used when compiling the Allocate op, and may be dynamic
+            // migraphx::shape compile_shape = migraphx::shape(s0.type(),{s0.dyn_dims().front(),
+            // {sample_size, sample_size} } );
+            migraphx::shape compile_shape =
+                migraphx::shape(s0.type(), {s0.dyn_dims().front(), {sample_size, sample_size}});
 
-        std::uniform_real_distribution<> dis(0.0, 1.0);
-        size_t batch_size = args[0]->get_shape().lens().front();
-        migraphx::shape dist_shape{migraphx::shape::float_type, {batch_size, sample_size}};
+            // Allocate on-device storage for the random values
+            // TODO:  this creates data type half_type if "buf_type" is not specified
+            auto alloc = info.add_instruction(
+                migraphx::make_op("allocate",
+                                  {{"shape", to_value(compile_shape)}, {"buf_type", s0.type()}}),
```

Comment:
You don't need to set `buf_type` if the `shape` attribute is set since `allocate` will use the type from `shape`. I can create a small update to change buf_type to default to nothing and add a check for it if the `shape` attribute is not set.

Reply:
Do you want to commit that to this PR, or I do it?  It won't pass tests without the fix.

Reply:
Working on it and an enhancement

Reply:
This is now merged.